WifiQueue.class.nut

```diff
@@ -1,170 +1,209 @@
-		_onFail = callback;
-	}
-
-
-	//-----------------------
-	function onDisconnect(callback) {
-		_onDisconnect = callback;
-	}
-
-	/*------------- PRIVATE FUNCTIONS -------------*/
-
-	//-----------------------
-	// Callback to run when device connects
-	function didConnect() {
-
-		_connecting = false;
-		_logs.log("Successfully connected to network: " + imp.getssid());
-
-		if (_onConnect) _onConnect();
-	}
-
-	//-----------------------
-	// Callback to run when device attempts and fails to connect
-	function didFail() {
-
-		_connecting = false;
-		_logs.log("Could not connect to network: " + imp.getssid());
-
-		if (_warDriving) {
-			if (_wifiList == null || _wifiList.len() == 0 || _currentNetwork == (_wifiList.len() - 1)) {
-				// Time to wardrive!
-				_wifiList = [];
-				_currentNetwork = 0;
-				local wifis = imp.scanwifinetworks();
-				foreach (wifi in wifis) {
-					if (wifi.open) {
-						_wifiList.push({"ssid": wifi.ssid, "pw": ""})
-					}
-				}
-			}
-		} else if (_currentNetwork == (_wifiList.len() - 1)) {
-			_logs.error("Failed to connect to any network");
-		}
-
-		local oldSsid = imp.getssid();
-
-		// Try the next network
-		if (_wifiList.len() > 0) {
-			_currentNetwork = _currentNetwork == null ? 0 : (_currentNetwork + 1) % _wifiList.len();
-			local network = _wifiList[_currentNetwork];
-			if ("ssid" in network && network.ssid.len() > 0 && "pw" in network) {
-				imp.setwificonfiguration(network.ssid, network.pw);
-			} else {
-				_logs.error("Skipping invalid wifi list entry " + _currentNetwork);
-			}
-		} else {
-			_logs.error("No networks to try");
-		}
-
-		// Throw the event
-		if (_onFail) _onFail(oldSsid);
-		if (!_connecting) {
-			connect();
-		}
-
-	}
-
-
-	//-----------------------
-	function didDisconnect(expected) {
-		_connecting = false;
-		if (_onDisconnect) _onDisconnect(expected);
-		if (!expected && !_connecting) {
-			connect();
-		}
-	}
+class WifiQueue {
+    static version = [3, 0, 0];
+
+    // Connection manager instance
+    _cm = null;
+    // List of wifi networks to try
+    _wifis = null;
+    // Whether we are currently trying to connect to a network
+    _connecting = false;
+    // Whether we should try networks that are not visible (because they may be hidden)
+    _hidden = null;
+    // Whether we should try open wifi networks (that are not in the list we were given)
+    _open = null;
+    // Whether we have scanned for open wifi networks to initiate warDriving yet
+    _warDriving = false;
+    // Logger object
+    _logger = null;
+
+    // Callbacks
+    _onFail = null;
+    _onConnect = null;
+    _onDisconnect = null;
+
+    // -----------------------
+    // `cm` is a connection manager instance
+    // `wifis` is an [ { ssid, pw } ]
+    // `flags` is a bit-array of boolean options`
+    // `logger` is a logger object with `.log()` and `.error()` methods
+    constructor(cm, flags = null, logger = server) {
+        _cm = cm;
+        _logger = logger;
```

Comment:
You shouldn't log unless the user has set a debug flag and if they have the logs should be prepended with: "WifiQueue:"